### ISSUE TITLE
Caching of tuning information for AutoGTP

### ISSUE THREAD

**Issue by wctgit** [gcp/leela-zero#713]
I noticed there's a file 'leelaz_opencl_tuning'. It looks like it's a generated file, perhaps from when leelaz.exe is run? However, it does not appear to be generated when autogtp.exe is run, and tuning appears to occur each and every time Auto GTP is run, regardless of whether leelaz_opencl_tuning is present or not.

Since tuning is annoyingly long, please consider caching the tuning results for AutoGTP.exe as well as leelaz.exe.

**Comment by Ttl** [comment]
The tuning is already reused. It should be run only once and the same tuning reused when autogtp is started the next time. If that doesn't happen for you then leelaz for some reason can't either find the tuning file or doesn't have permission to write it.

**Comment by wctgit** [comment]
Every time I run autogtp, tuning takes about 5 seconds. There is no indication whether cached tuning was used or whether it was re-performed. I suspect the latter, but have no way of knowing. No cache file is generated when I remove the leelaz_opencl_tuning. There should not be any issues with permissions, as it is able to create other files such as the network gz files. Without info/warning/error messages, I have no way of knowing. Perhaps some diagnostic feedback output should be added to the startup procedure.

**Comment by gcp** [comment]
Leelaz stored the tuning and AutoGTP just calls leelaz, which reuses the stored tuning.

Together with #712 it seems the problem is that your system is seriously fucked up.

**Comment by roy7** [comment]
When Leela runs, if it's using existing tuning, you'll see a message during Leela's startup "Loaded existing SGEMM tuning". Maybe try running autogtp with admin privileges? Perhaps the programs aren't allowed to write to their directory.

**Comment by wctgit** [comment]
> Together with #712 it seems the problem is that your system is seriously fucked up.

Actually, the problem is with gzip.exe, as identified in the comments in #712. Well, that and the confusing message saying "Starting tuning process, please wait...", when actually it starts by downloading the network first, which was what actually caused the delay, rather than the tuning process. Had it said something like "Fetching neural network, please wait...", I wouldn't have made this report.

**Comment by marcocalignano** [comment]
Downloading the network is part of the tuning process! 
So you want autogtp to write:

```
Starting tuning process, please wait...
Fetching job ...
Fetching neural network, please wait...
Decompressing neural network, please wait...
Decompression done.
Fetching neural network done.
Fetching job done.
.......
Tuning done.
```
so you can just tell us that you have a broken gzip program (thing that we cannot test since our is working?) I got that the problem was with your system immidiately when you said that the network was delete. Autogtp does not deletes networks ever so it must have been your system.

**Comment by wctgit** [comment]
> Downloading the network is part of the tuning process!

Isn't it more accurate to say that the tuning process depends on having a local network, which is dependent on the downloading process? Because you can do other things with the network aside from tuning. Tuning is done after downloading.

>So you want autogtp to write:

What seems important to me is whether the output is confusing or not. I wrote this issue because the output said 'tuning' and there was a long delay. But the delay was due to downloading. Tuning wasn't even getting started yet. That was not clear from the output. Hence I associated the long delay with 'tuning'.

> so you can just tell us that you have a broken gzip program (thing that we cannot test since our is working?)

The only gzip.exe I have is the one from the Leela Zero distribution, so if it's broken, then Leela Zero distribution contains a broken gzip.exe. You can test it: You could recreate the same simple folder structure on a Windows machine and try the steps I wrote in issue #712.

> I got that the problem was with your system immidiately when you said that the network was delete. Autogtp does not deletes networks ever so it must have been your system.

I said that it deletes the .gz file, not the network file with no filename extension. There's nothing special about my system. It's a typical Windows laptop with hardly anything installed on it. In fact, I found another issue by someone else who reported the same bug with gzip: #342. It's a quirk/bug in gzip.exe (which is the one distributed by Leela Zero release), not my specific system.

**Comment by marcocalignano** [comment]
> You could recreate the same simple folder structure on a Windows machine and try the steps I wrote in issue

No I cannot I do not own any Windows system.

**Comment by wctgit** [comment]
You said, "we cannot test," as plural. I was responding with the plural form of 'you', meaning anyone involved in the project (who has a Windows machine available).
